### PR TITLE
AG-208: Feature enrichment for XAResourceWrapper

### PR DESCRIPTION
**Pull request by graben**
@barreiro : Could you plz review?

### CODE REVIEW

**Comment by barreiro** [comment]
thanks @graben. I'm writing a couple more test cases and then I'll be able to merge this PR.

**Review by barreiro — CHANGES_REQUESTED** [review]

File: agroal-narayana/src/main/java/io/agroal/narayana/FirstResourceBaseXAResource.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// You may not use this file except in compliance with the Apache License, Version 2.0.
+
+package io.agroal.narayana;
+
+import io.agroal.api.transaction.TransactionAware;
+import org.jboss.tm.FirstResource;
+
+import javax.transaction.xa.XAResource;
+
+public class FirstResourceBaseXAResource extends BaseXAResource implements FirstResource {
```
[the remark this answers is not in the document]
Comment (graben) [reply agroal-narayana/src/main/java/io/agroal/narayana/FirstResourceBaseXAResource.java:11]: Look at the linked issue from narayana-spring-boot [#85](https://github.com/snowdrop/narayana-spring-boot/issues/85). I think firstresource is the only way to solve this problem with Narayana.

File: agroal-narayana/pom.xml [hunk 2]
```diff
@@ -20,5 +20,10 @@
             <artifactId>jboss-transaction-spi</artifactId>
             <version>${version.org.jboss.jboss-transaction-spi}</version>
         </dependency>
+        <dependency>
+            <groupId>org.jboss.narayana.jta</groupId>
+            <artifactId>narayana-jta</artifactId>
+            <version>${version.org.jboss.narayana}</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (graben) [reply agroal-narayana/pom.xml]: Done

File: agroal-narayana/src/main/java/io/agroal/narayana/NarayanaTransactionIntegration.java [hunk 3]
```diff
@@ -60,10 +62,19 @@ public NarayanaTransactionIntegration(TransactionManager transactionManager, Tra
...
+    public NarayanaTransactionIntegration(TransactionManager transactionManager, TransactionSynchronizationRegistry transactionSynchronizationRegistry, String jndiName, boolean connectable, boolean firstResource) {
+        this( transactionManager, transactionSynchronizationRegistry, jndiName, connectable, firstResource, null );
+    }
+
+    public NarayanaTransactionIntegration(TransactionManager transactionManager, TransactionSynchronizationRegistry transactionSynchronizationRegistry, String jndiName, boolean connectable, boolean firstResource, XAResourceRecoveryRegistry recoveryRegistry) {
         this.transactionManager = transactionManager;
         this.transactionSynchronizationRegistry = transactionSynchronizationRegistry;
         this.jndiName = jndiName;
         this.connectable = connectable;
+        this.firstResource = firstResource;
```
[the remark this answers is not in the document]
Comment (graben) [reply agroal-narayana/src/main/java/io/agroal/narayana/NarayanaTransactionIntegration.java:80]: Done

File: agroal-narayana/src/main/java/io/agroal/narayana/NarayanaTransactionIntegration.java [hunk 4]
```diff
@@ -93,7 +104,11 @@ public void associate(TransactionAware transactionAware, XAResource xaResource)
 
                     XAResource xaResourceToEnlist;
                     if ( xaResource != null ) {
-                        xaResourceToEnlist = new BaseXAResource( transactionAware, xaResource, jndiName );
+                        if ( firstResource ) {
+                            xaResourceToEnlist = new FirstResourceBaseXAResource( transactionAware, xaResource, jndiName );
+                        } else {
+                            xaResourceToEnlist = new BaseXAResource( transactionAware, xaResource, jndiName );
+                        }
```
[the remark this answers is not in the document]
Comment (graben) [reply agroal-narayana/src/main/java/io/agroal/narayana/NarayanaTransactionIntegration.java]: Done

File: agroal-narayana/src/main/java/io/agroal/narayana/ConnectableLocalXAResource.java [hunk 5]
```diff
@@ -5,12 +5,11 @@
 
 import io.agroal.api.transaction.TransactionAware;
 import org.jboss.tm.ConnectableResource;
-import org.jboss.tm.LastResource;
 
 /**
  * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
  */
-public class ConnectableLocalXAResource extends LocalXAResource implements ConnectableResource, LastResource {
+public class ConnectableLocalXAResource extends LocalXAResource implements ConnectableResource {
```
[the remark this answers is not in the document]
Comment (graben) [reply agroal-narayana/src/main/java/io/agroal/narayana/ConnectableLocalXAResource.java]: Done

File: agroal-narayana/src/main/java/io/agroal/narayana/LocalXAResource.java [hunk 6]
```diff
@@ -19,7 +21,7 @@
  * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
  * @author <a href="jesper.pedersen@redhat.com">Jesper Pedersen</a>
  */
-public class LocalXAResource implements XAResourceWrapper {
+public class LocalXAResource implements XAResourceWrapper, LastResource, LastResourceCommitOptimisation {
```
[the remark this answers is not in the document]
Comment (graben) [reply agroal-narayana/src/main/java/io/agroal/narayana/LocalXAResource.java]: Revert done

**Comment by graben** [comment]
@barreiro : Feedback added to PR

**Review by barreiro — APPROVED** [review]

**Comment by graben** [comment]
Hi @barreiro, I'm just thinking about CommitMarkableResource wouldn't really work well with Spring-Boot as DataSource isn't being bind to JNDI. I think I will work on another improvement for this.

**Comment by barreiro** [comment]
@graben that would be an improvement to either Narayana or Spring-Boot, no ?? 

The [`RecoveryModule`](https://github.com/jbosstm/narayana/blob/7.0.0.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore//7.0.0.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/CommitMarkableResourceRecordRecoveryModule.java#L210-L212CommitMarkableResourceRecordRecoveryModule.java#L210-L212) always fetch the datasource from JNDI.

By the way, I'm not sure Quarkus has everything need for CMR either.

**Comment by graben** [comment]
@barreiro : It'll be an improvement to the Spring-Boot starter only. I'm just coding a bit in a WIP branch [AG-xxx](https://github.com/graben/agroal/tree/AG-xxx) I'll submit PR once ready.

**Comment by barreiro** [comment]
@graben thanks for letting me know. I have left some comments.

**Pull request merged**